shared/src/permissions.rs

```diff
@@ -0,0 +1,157 @@
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+use strum::IntoEnumIterator;
+
+#[derive(Clone, Copy, Hash, Eq, PartialEq, Debug, Serialize, Deserialize)]
+pub enum Permission {
+    Account(AccountPermission),
+    Universe(GenericPermission),
+}
+
+#[derive(
+    Clone,
+    Copy,
+    Hash,
+    Eq,
+    PartialEq,
+    Debug,
+    Serialize,
+    Deserialize,
+    strum_macros::EnumString,
+    strum_macros::ToString,
+    strum_macros::EnumIter,
+)]
+pub enum AccountPermission {
+    PermanentBan,
+    TemporaryBan,
+    List,
+    View,
+}
+
+impl ToString for Permission {
+    fn to_string(&self) -> String {
+        let (service, permission) = match self {
+            Permission::Account(perm) => (Service::Account, perm.to_string()),
+            Permission::Universe(perm) => (Service::Universe, perm.to_string()),
+        };
+
+        format!("{}({})", service.to_string(), permission)
+    }
+}
+
+impl FromStr for Permission {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut service = String::new();
+        let mut permission = String::new();
+        enum State {
+            InService,
+            InPermission,
+            AtEnd,
+        }
+        let mut state = State::InService;
+
+        for c in s.chars() {
+            match c {
+                '(' => {
+                    state = State::InPermission;
+                }
+                ')' => {
+                    state = State::AtEnd;
+                }
+                _ => match state {
+                    State::InService => service.push(c),
+                    State::InPermission => permission.push(c),
+                    State::AtEnd => anyhow::bail!("junk found after close paren"),
+                },
+            }
+        }
+
+        let service = Service::from_str(&service)?;
+
+        let permission = match service {
+            Service::Account => Permission::Account(AccountPermission::from_str(&permission)?),
+            Service::Universe => Permission::Universe(GenericPermission::from_str(&permission)?),
+        };
+
+        Ok(permission)
+    }
+}
```

Comment:
Do we even need any of these `str` implementations if we just save permissions in a (stable) binary format?

Reply:
I'm serializing these values into Postgres, that's why I need these values. Serde is being used once these are put in the protocol.

Reply:
Am I getting this right: this code going to go away when you have done that?
Why can't you go though CBOR directly here, to `Serialize` and `Deserialize`?

Reply:
I don't want to lose the ability to query the values in Postgres -- CBOR is binary, JSON feels overkill, these strings work great.

Plus, I can use these in the CLI (starting to implement it right now), e.g.: `cosmicverge-server permission-groups add-permission mygroup "Account(View)"`

Reply:
Sounds good.